Add dropout removal pass.

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38254 Enabled dropout removal pass in mobile optimizer.
* **#38253 Add dropout removal pass.**

This pass removes dropout and dropout_ nodes when training is false. It
requires to have run freeze_module pass which does both inlining and constant
propagation, without which training variable remains as attribute instead of
constant.

Differential Revision: [D21505863](https://our.internmc.facebook.com/intern/diff/D21505863/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21505863/)!

kimishpatel committed:
Add dropout removal pass.

This pass removes dropout and dropout_ nodes when training is false. It
requires to have run freeze_module pass which does both inlining and constant
propagation, without which training variable remains as attribute instead of
constant.

Differential Revision: [D21505863](https://our.internmc.facebook.com/intern/diff/D21505863/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21505863/)!

[ghstack-poisoned]

jamesr66a commented on torch/csrc/jit/passes/remove_dropout.cpp:
```diff
@@ -0,0 +1,54 @@
+  TORCH_INTERNAL_ASSERT(inputs.size() == 3);
+  const Value* training_input = inputs[2];
+  auto optional_ivalue = toIValue(training_input);
+  TORCH_INTERNAL_ASSERT(optional_ivalue.has_value());
```
This is going to crash the program when `train` isn't a constant, isn't it? e.g.

```
import torch
import torch.nn.functional as F

class TestDropout(torch.nn.Module):
    def forward(self, x, training : bool):
        return torch.dropout(x, 0.5, training)

dropout_training = torch.jit.script(TestDropout())

print(dropout_training.graph)
```

```
graph(%self : __torch__.TestDropout,
      %x.1 : Tensor,
      %training.1 : bool):
  %4 : float = prim::Constant[value=0.5]() # dropout.py:6:32
  %6 : Tensor = aten::dropout(%x.1, %4, %training.1) # dropout.py:6:15
  return (%6)
```

kimishpatel replied:
Yes, indeed. Thanks for pointing this out. Will put guard around it.

kimishpatel replied:
This was intended to be deployed post freezing, but still, this has to be self-contained so will fix this.

kimishpatel replied:
Probably the best course is to not remove such dropout and just leave a warning that it was not removed.

kimishpatel committed:
Update on "Add dropout removal pass."


This pass removes dropout and dropout_ nodes when training is false. It
requires to have run freeze_module pass which does both inlining and constant
propagation, without which training variable remains as attribute instead of
constant.

Differential Revision: [D21505863](https://our.internmc.facebook.com/intern/diff/D21505863/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21505863/)!

[ghstack-poisoned]

kimishpatel committed:
Update on "Add dropout removal pass."


This pass removes dropout and dropout_ nodes when training is false. It
requires to have run freeze_module pass which does both inlining and constant
propagation, without which training variable remains as attribute instead of
constant.

Differential Revision: [D21505863](https://our.internmc.facebook.com/intern/diff/D21505863/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21505863/)!

[ghstack-poisoned]
